File: lintcode/python/0132_word_search_II.py
#!/usr/bin/python -t

# trie

class trieNode:
    def __init__(self):
        self.children = {}
        self.is_word = False
        self.word = None
        
class trie:
    def __init__(self):
        self.root = trieNode()
        
    def insert(self, word):
        cur = self.root
        
        for c in word:
            if c not in cur.children:
                cur.children[c] = trieNode()
            
            cur = cur.children[c]
        cur.is_word = True
        cur.word = word
        
        return
    
    def find(self, word):
        cur = self.root
        
        for c in word:
            cur = cur.children.get(c)
            if cur == None:
                return None
        return cur
        

class Solution:
    """
    @param board: A list of lists of character
    @param words: A list of string
    @return: A list of string
    """
    def wordSearchII(self, board, words):
        # write your code here
        m = len(board)
        if m == 0:
            return []
        n = len(board[0])
        if n == 0:
            return []
            
        t = trie()
        
        for w in words:
            t.insert(w)
            
        ret = set()
        #v = set()
        
        for i in range(m):
            for j in range(n):
                self.search(board, i, j, t.root.children.get(board[i][j]), set([(i,j)]), ret)
                #v.remove((i, j))
                
        return list(ret)
        
    def search(self, board, x, y, node, v, ret):
        if node == None:
            return
        if node.is_word:
            ret.add(node.word)
            
        dx = [1, -1, 0, 0]
        dy = [0, 0, 1, -1]
        
        for i in range(4):
            cx = dx[i] + x
            cy = dy[i] + y
            
            if not self.valid(board, cx, cy, v):
                continue
            v.add((cx, cy))
            self.search(board, cx, cy, node.children.get(board[cx][cy]), v, ret)
            v.remove((cx, cy))
            
        return
    
    def valid(self, board, x, y, v):
        m, n = len(board), len(board[0])
        
        #print v
        if x < 0 or y < 0 or x >= m or y >= n or (x,y) in v:
            return False
        return True
        
            

# dfs
class Solution:
    """
    @param board: A list of lists of character
    @param words: A list of string
    @return: A list of string
    """
    def wordSearchII(self, board, words):
        # write your code here
        if len(words) == 0:
            return []
        maxLen = max(len(word) for word in words)
        ret = []
        path = ""
        v = set()
        prefix_set = set()
        for word in words:
            for i in range(len(word)):
                prefix_set.add(word[:i + 1])

        for i in range(len(board)):
            for j in range(len(board[0])):
                v.add((i, j))
                self.dfs(board, words, prefix_set, maxLen, i, j, board[i][j], v, ret)
                v.remove((i, j))

        return ret

    def dfs(self, board, words, prefix_set, maxLen, x, y, path, v, ret):
        if path not in prefix_set:
            return
        if len(path) > maxLen:
            return
        if path in words:
            if path not in ret:
                ret.append(path)
        
        dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        for i in range(len(dirs)):
            nx = x + dirs[i][0]
            ny = y + dirs[i][1]

            if 0 <= nx < len(board) and 0 <= ny < len(board[0]) and (nx, ny) not in v:
                v.add((nx, ny))
                self.dfs(board, words, prefix_set, maxLen, nx, ny, path+board[nx][ny], v, ret)
                v.remove((nx, ny))

        return

if __name__ == '__main__':
    s = Solution()
    a =["abce","sfcs","adee"]
    b= ["see","se"]
    print(s.wordSearchII(a,b))

File: lintcode/golang/0132_word_search_II.go
package main

// trie

import (
	"fmt"
	"sort"
)

type trie struct {
	children map[byte]*trie
	isWord   bool
	word     string
}

var root *trie

func newTrie() *trie {
	return &trie{
		children: make(map[byte]*trie),
		isWord:   false,
	}
}

func (t *trie) insert(word string) {
	cur := root
	for idx := range word {
		if nextCur, ok := cur.children[word[idx]]; !ok {
			nextCur = newTrie()
			cur.children[word[idx]] = nextCur
		}
		cur = cur.children[word[idx]]
	}
	cur.isWord = true
	cur.word = word
	return
}

/**
 * @param board: A list of lists of character
 * @param words: A list of string
 * @return: A list of string
 */
func wordSearchII(board [][]byte, words []string) []string {
	// write your code here
	m := len(board)
	if m == 0 {
		return []string{}
	}
	n := len(board[0])
	if n == 0 {
		return []string{}
	}

	root = newTrie()

	for idx := range words {
		root.insert(words[idx])
	}

	ret := []string{}
	visited := make([][]bool, m, m)
	for i := range visited {
		visited[i] = make([]bool, n, n)
	}
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			//visited[i][j] = true
			search(board, i, j, root, visited, &ret)
			//visited[i][j] = false
		}
	}
	sort.Strings(ret)
	return ret
}

func search(board [][]byte, x, y int, node *trie, visited [][]bool, ret *[]string) {
	fmt.Printf("searching %v, %v, %v, %v, %v\n", x, y, string(board[x][y]), visited, node.children)
	if _, ok := node.children[board[x][y]]; !ok {
		return
	}

	next := node.children[board[x][y]]
	if next.isWord {
		*ret = append(*ret, next.word)
		next.isWord = false
		//return
	}

	visited[x][y] = true
	direcs := []struct {
		dx int
		dy int
	}{{-1, 0}, {0, -1}, {0, 1}, {1, 0}}
	for i := 0; i < 4; i++ {
		nx := x + direcs[i].dx
		ny := y + direcs[i].dy

		if valid(board, nx, ny, visited) {
			//visited[nx][ny] = true
			search(board, nx, ny, next, visited, ret)
			//visited[nx][ny] = false
		}
	}
	visited[x][y] = false

	return
}

func valid(board [][]byte, x int, y int, visited [][]bool) bool {
	m := len(board)
	n := len(board[0])
	if x < 0 || x >= m || y < 0 || y >= n || visited[x][y] {
		return false
	}
	return true
}

func main() {
	/*
		b := make([][]byte, 3)
		b[0] = []byte("doaf")
		b[1] = []byte("agai")
		b[2] = []byte("dcan")

		w := []string{"dog", "dad", "dgdg", "can", "again"}
	*/
	/*
		b := make([][]byte, 3)
		b[0] = []byte("abce")
		b[1] = []byte("sfcs")
		b[2] = []byte("adee")

		w := []string{"see", "se"}
	*/
	b := make([][]byte, 3)
	b[0] = []byte("abce")
	b[1] = []byte("sfcs")
	b[2] = []byte("adee")

	w := []string{"as", "ab", "cf", "da", "ee", "e", "adee", "eeda"}

	fmt.Println(wordSearchII(b, w))
}



// trie

/**
 * @param board: A list of lists of character
 * @param words: A list of string
 * @return: A list of string
 */
import "sort"

var directions = [][]int{
	[]int{1, 0},
	[]int{0, 1},
	[]int{0, -1},
	[]int{-1, 0},
}

func wordSearchII(board [][]byte, words []string) []string {
	m, n := len(board), 0
	if m == 0 {
		return []string{}
	}
	n = len(board[0])

	trie := NewTrie()
	for _, word := range words {
		trie.Add(word)
	}

	visited := make([][]bool, m, m)
	for i := range visited {
		visited[i] = make([]bool, n, n)
	}

	result := make([]string, 0)
	for i := range board {
		for j := range board[i] {
			dfs(board, i, j, visited, trie.Root, &result)
		}
	}
	sort.Strings(result)
	return result
}

func dfs(board [][]byte, x, y int, visited [][]bool, node *TrieNode, res *[]string) {
	child := node.Children[board[x][y]-'a']
	if child == nil {
		return
	}

	if child.Word != nil {
		*res = append(*res, *child.Word)
		child.Word = nil
	}

	visited[x][y] = true
	for _, direction := range directions {
		newX, newY := x+direction[0], y+direction[1]
		if newX < 0 || newY < 0 || newX >= len(board) || newY >= len(board[0]) {
			continue
		}
		if visited[newX][newY] {
			continue
		}
		dfs(board, newX, newY, visited, child, res)
	}
	visited[x][y] = false
}

type TrieNode struct {
	Children []*TrieNode
	Word     *string
}

func NewTrieNode() *TrieNode {
	return &TrieNode{Children: make([]*TrieNode, 26, 26)}
}

type Trie struct {
	Root *TrieNode
}

func NewTrie() *Trie {
	return &Trie{Root: NewTrieNode()}
}

func (t *Trie) Add(word string) {
	node := t.Root
	for _, c := range word {
		idx := c - 'a'
		if node.Children[idx] == nil {
			node.Children[idx] = NewTrieNode()
		}
		node = node.Children[idx]
	}
	node.Word = &word
}
